### PR TITLE
Gemm reduce examples int4/int8/fp32/bf16

### PR DESCRIPTION
**Pull request by aosewski**

### CODE REVIEW

**Comment by aosewski** [comment]
@asroy @rocking5566  I found that int8 example for `mean_meansquare` is working only if `ReduceAccDataType` and `R[0|1]DataType` is F32 and not int32. I had an impression that `AtomicAdd` should work also for `int32_t`: 

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/utility/reduction_operator.hpp#L53-L56

https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/utility/dynamic_buffer.hpp#L309

**Comment by rocking5566** [comment]
> @asroy @rocking5566 I found that int8 example for `mean_meansquare` is working only if `ReduceAccDataType` and `R[0|1]DataType` is F32 and not int32. I had an impression that `AtomicAdd` should work also for `int32_t`:
> 
> https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/utility/reduction_operator.hpp#L53-L56
> 
> https://github.com/ROCmSoftwarePlatform/composable_kernel/blob/develop/include/ck/utility/dynamic_buffer.hpp#L309

Please test this branch I give you before.
https://github.com/ROCmSoftwarePlatform/composable_kernel/commit/1fc39f634f5c0bdde2275ed9c13e1f0e8e8feeaa
You need to remove div first.

File: example/16_gemm_multi_d_multi_reduces/gemm_mean_meansquare_xdl_fp32.cpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+using ELayout = Row;
+
+// Elementwise op
+using Square       = ck::tensor_operation::element_wise::UnarySquare;
+using Div          = ck::tensor_operation::element_wise::UnaryDivide;
+using AElementOp   = PassThrough;
+using BElementOp   = PassThrough;
+using CDEElementOp = PassThrough;
+using QsElementOp  = ck::Tuple<PassThrough, Square>;
+using RsElementOp  = ck::Tuple<Div, Div>;
```
Comment (rocking5566) [inline example/16_gemm_multi_d_multi_reduces/gemm_mean_meansquare_xdl_fp32.cpp:34]: div might fail in integer

File: example/16_gemm_multi_d_multi_reduces/gemm_reduce_xdl_common.hpp [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+        a_device_buf.ToDevice(a_m_k.mData.data());
+        b_device_buf.ToDevice(b_k_n.mData.data());
+    }
+#endif // CK_EXPERIMENTAL_BIT_INT_EXTENSION_INT4
+
+        auto a_element_op = AElementOp{};
+    auto b_element_op   = BElementOp{};
+    auto cde_element_op = CDEElementOp{};
+    auto qs_element_op  = QsElementOp{};
+    auto rs_element_op  = RsElementOp{N, N};
```
Comment (rocking5566) [inline example/16_gemm_multi_d_multi_reduces/gemm_reduce_xdl_common.hpp:363]: try not to div

**Review by asroy — APPROVED** [review]

File: example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+            r0_m_host(m) = ck::type_convert<R0DataType>(reduce0_acc);
+            r1_m_host(m) = ck::type_convert<R1DataType>(reduce1_acc);
+        }
+        e_device_buf.FromDevice(e_m_n.mData.data());
+
+        Tensor<INT32> e_m_n_converted(e_m_n_host);
+
+        pass = ck::utils::check_err(
+            e_m_n_converted.mData, e_m_n_host.mData, "Error: Incorrect results c", 1e-2, 1e-2);
```
Comment (rocking5566) [inline example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp:267]: you should compare (int8)e_m_n_host with e_m_n.
However, you compare e_m_n_host with (int32)e_m_n_host here.

File: example/16_gemm_multi_d_multi_reduces/gemm_max_xdl_int8.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+#include "ck/library/reference_tensor_operation/cpu/reference_gemm.hpp"
+#include "ck/tensor_operation/gpu/device/device_gemm_multiple_d_multiple_r_xdl_cshuffle.hpp"
+#include "ck/tensor_operation/gpu/device/gemm_specialization.hpp"
+
+using ADataType         = INT8;
+using BDataType         = INT8;
+using GemmAccDataType   = INT32;
+using CShuffleDataType  = INT32;
+using DsDataType        = ck::Tuple<>;
+using EDataType         = INT32;
```
Comment (rocking5566) [inline example/16_gemm_multi_d_multi_reduces/gemm_max_xdl_int8.cpp:15]: Sync the style withgemm_add_addsquare_xdl_int8.cpp. using EDataType = INT8;

File: example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        if(pass)
+        {
+            std::cout << "Success!" << std::endl;
+        }
+    }
+
+    if(time_kernel)
+    {
+        float ave_time            = invoker.Run(argument, StreamConfig{nullptr, time_kernel});
+        std::size_t flop          = std::size_t(2) * M * N * K + M * 3 * N;
```
Comment (rocking5566) [inline example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp:286]: std::size_t(3)

File: example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp (hunk 3 above)
Comment (aosewski) [reply example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp]: fixed [88ce32a](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/368/commits/88ce32ac59516d2a5151e12e6995d5290685c83a)

File: example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp (hunk 5 above)
Comment (aosewski) [reply example/16_gemm_multi_d_multi_reduces/gemm_add_addsquare_xdl_int8.cpp]: done [bb46cf8](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/368/commits/bb46cf807fdb50e5eda14e13100e13a19545e8ed)

File: example/16_gemm_multi_d_multi_reduces/gemm_max_xdl_int8.cpp (hunk 4 above)
Comment (aosewski) [reply example/16_gemm_multi_d_multi_reduces/gemm_max_xdl_int8.cpp]: Done [88ce32a](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/368/commits/88ce32ac59516d2a5151e12e6995d5290685c83a)

**Review by asroy — APPROVED** [review]

**Pull request merged**